kroxylicious/src/main/java/io/kroxylicious/proxy/internal/codec/DecodePredicate.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.internal.codec;
+
+import java.util.Arrays;
+
+import org.apache.kafka.common.protocol.ApiKeys;
+
+import io.kroxylicious.proxy.filter.KrpcFilter;
+
+/**
+ * Encapsulates decisions about whether requests and responses should be
+ * fully deserialized into POJOs, or passed through as byte buffers with
+ * minimal deserialization.
+ *
+ * The actual decision can depend on which filters are in use, which can depend on
+ * who the authorized user or, or which back-end cluster they're connected to.
+ */
+public interface DecodePredicate {
+    public static DecodePredicate forFilters(KrpcFilter... filters) {
+        return new DecodePredicate() {
+            @Override
+            public boolean shouldDecodeResponse(ApiKeys apiKey, short apiVersion) {
+                for (var filter : filters) {
+                    if (filter.shouldDeserializeResponse(apiKey, apiVersion)) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            @Override
+            public boolean shouldDecodeRequest(ApiKeys apiKey, short apiVersion) {
+                for (var filter : filters) {
+                    if (filter.shouldDeserializeRequest(apiKey, apiVersion)) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            @Override
+            public String toString() {
+                return "DecodePredicate$forFilters{" + Arrays.toString(filters) + "}";
+            }
+        };
+    }
+
+    public boolean shouldDecodeRequest(ApiKeys apiKey, short apiVersion);
```

Comment:
do we need both `shouldDecodeRequest` and `shoulcDecodeResponse` as separate methods or would a single `shouldDecode`  suffice.

Reply:
I think they should remain separate. You might want to intercept only Produce requests, for example, so why incur the expense of having to deserialize (and pass through the pipeline) every Produce response?

Reply:
~If you only wanted to deserialize produce requests you would just return `false` for the response. 

I'm happy to keep them separate, just wondering if there is really a conceptually different. ~

I think I've just twigged why they are separate! The API Key is the same in both directions thus request and response can only be inferred from the method call. At which point yes keep them separate!